python/plugins/processing/algs/qgis/RasterCalculator.py

```diff
@@ -213,3 +225,69 @@ def processBeforeAddingToModeler(self, algorithm, model):
                         values.append(ValueFromOutput(alg.modeler_name, out.name))
 
         algorithm.params[self.LAYERS] = values
+
+    def mappedNameToLayer(self, lyr, expression, layersDict, context):
+        '''Try to identify if a real layer is mapped in the expression with a symbolic name.'''
+
+        nameToMap = lyr.name()
+
+        # get last scope of the expressionContext because should be that related
+        # with mapped variables
+        # The scope name should be "algorithm_inputs"
+        expContextLastScope = context.expressionContext().lastScope()
+
+        # check for layers directly added in the expression
+        if (nameToMap + "@") in expression:
+            layersDict[nameToMap] = lyr
+
+        # check for the layers that are mapped as input in a model
+        #  to do this check in the latest scope all passed variables
+        # to look for a variable that is a layer or a string filename ç
+        # to a layer
+        varId = None
+        varDescription = None
+
+        for varName in expContextLastScope.variableNames():
+
+            layer = expContextLastScope.variable(varName)
+
+            if not isinstance(layer, str) and not isinstance(layer, QgsMapLayer):
+                continue
+
+            if isinstance(layer, QgsMapLayer) and nameToMap not in layer.source():
```

Comment:
I don't fully understand the logic here. Earlier you're comparing against the layer name, but here against the layer source?

Reply:
you are right